### PR TITLE
refactor(common): deduplicate chunk operation implementations

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Move `StreamChunkBuilder` to `common` crate.
- Implement several interfaces with `ChunkBuilder` to avoid code duplication, as mentioned in #13903 

This PR intentionally left the signatures untouched. Planned to make them consistent between `DataChunk` and `StreamChunk` in future works.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/array/stream_chunk.rs
Comment (fuyufjh) [inline src/common/src/array/stream_chunk.rs:1]: May remove/update the `TODO` at L128.

https://github.com/risingwavelabs/risingwave/blob/51c76aaa13ac4e81b61d37a07be1ac7afc6631a9/src/common/src/array/stream_chunk.rs#L127-L129

**Review by fuyufjh — APPROVED** [review]
LGTM. Didn't go through the details since the test case should cover them well.

File: src/common/src/array/data_chunk.rs [hunk 1]
```diff
@@ -95,23 +96,17 @@ impl DataChunk {
...
 
     /// Build a `DataChunk` with rows.
+    ///
+    /// Panics if the `rows` is empty.
     pub fn from_rows(rows: &[impl Row], data_types: &[DataType]) -> Self {
-        let mut array_builders = data_types
-            .iter()
-            .map(|data_type| data_type.create_array_builder(1))
-            .collect::<Vec<_>>();
+        let mut builder = DataChunkBuilder::new(data_types.to_vec(), rows.len() + 1);
```
Comment (TennyZhuang) [inline src/common/src/array/data_chunk.rs:102]: Why `+1` here?

Comment (BugenZhao) [reply src/common/src/array/data_chunk.rs:105]: For code simplicity, we can consume the builder after the loop. Otherwise, it'll output on the last cycle. Not a big deal.

Comment (kwannoel) [reply src/common/src/array/data_chunk.rs:105]:
```suggestion
        // `append_one_row` will cause the builder to finish immediately once capacity is met.
        // Hence, we allocate an extra row (rows.len() + 1), to avoid the builder finishing prematurely.
        // This just makes the code cleaner, since we can loop through all rows, and consume it finally.
        let mut builder = DataChunkBuilder::new(data_types.to_vec(), rows.len() + 1);
```

File: src/common/src/array/data_chunk.rs [hunk 2]
```diff
@@ -322,83 +320,24 @@ impl DataChunk {
...
-                    .map(|col_type| col_type.create_builder(new_chunk_require))
-                    .collect();
-
-                let data_chunk = DataChunk::new(new_columns, array_len);
-                new_chunks.push(data_chunk);
-
-                new_chunk_require = std::cmp::min(total_capacity, each_size_limit);
-                array_len = new_chunk_require;
+        for chunk in chunks {
+            for output in builder.append_chunk(chunk.clone()) {
```
Comment (kwannoel) [inline src/common/src/array/data_chunk.rs:332]:
```suggestion
            if let Some(output) = builder.append_chunk(chunk.clone()) {
```

nit: I prefer `if let` here, `for` seems to signal that this can happen multiple times, it's more general than `if let`.

File: src/common/src/array/stream_chunk.rs [hunk 3]
```diff
@@ -125,26 +126,20 @@ impl StreamChunk {
...
+            let none = builder.append_row(*op, row);
+            debug_assert!(none.is_none());
         }
 
-        let new_columns = array_builders
-            .into_iter()
-            .map(|builder| builder.finish().into())
-            .collect::<Vec<_>>();
-        StreamChunk::new(ops, new_columns)
+        builder.take().expect("empty chunk")
```
Comment (kwannoel) [inline src/common/src/array/stream_chunk.rs:142]: Prefer it to be type-safe, and return a None instead. But I think it will lead to large amount of refactoring, so we shouldn't refactor in this PR.

I guess subsequently we will refactor callers to use `StreamChunkBuilder` instead.

File: src/common/src/array/stream_chunk_builder.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            })
+            .map(Into::into)
+            .collect::<Vec<_>>();
+
+        Some(StreamChunk::new(
+            std::mem::replace(&mut self.ops, Vec::with_capacity(self.capacity)),
+            new_columns,
+        ))
+    }
+}
```
Comment (kwannoel) [inline src/common/src/array/stream_chunk_builder.rs:146]: Are there any diffs to review for this file? Seems like it's entirely migrated from `comm/builder.rs`

**Comment by kwannoel** [comment]
LGTM when tests are fixed. Probably the assertion for non-empty stream chunk broke some tests. Shall we permit returning empty stream chunk for now, to avoid breaking the function interfaces?

File: src/common/src/array/data_chunk.rs (hunk 1 above)
Comment (stdrc) [reply src/common/src/array/data_chunk.rs:105]: We may have something like `DataChunkBuilder::new_unlimited` to allow manually consuming chunks

File: src/common/src/array/data_chunk.rs [hunk 5]
```diff
@@ -95,23 +96,20 @@ impl DataChunk {
...
+            let none = builder.append_one_row(row);
+            debug_assert!(none.is_none());
         }
 
-        let new_columns = array_builders
-            .into_iter()
-            .map(|builder| builder.finish().into())
-            .collect::<Vec<_>>();
-        DataChunk::new(new_columns, rows.len())
+        builder.consume_all().expect("empty chunk")
```
Comment (stdrc) [inline src/common/src/array/data_chunk.rs:112]: It'd be better to write expect message like `builder should contain at least one row`.

File: src/common/src/array/data_chunk.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/array/data_chunk.rs:105]: Exactly! Actually I'm considering adding a new interface of `new_unlimited(initial_capacity: usize)` to decouple the output chunk size and memory reservation for array builders, so that the chunk size can be `usize::MAX`. Maybe in next PRs when we're going to unify the interfaces for `Stream` and `Data`.

File: src/common/src/array/data_chunk.rs (hunk 2 above)
Comment (BugenZhao) [reply src/common/src/array/data_chunk.rs:332]: Appending a large chunk may result in multiple smaller outputs. 😕 So the return type of `append_chunk` should be an iterator.

File: src/common/src/array/stream_chunk.rs (hunk 3 above)
Comment (BugenZhao) [reply src/common/src/array/stream_chunk.rs:142]: I have reviewed all use cases for these functions and found all of them correctly handle the empty case before calling this function. I guess the unit tests run into some edge cases and will check then.

File: src/common/src/array/stream_chunk_builder.rs (hunk 4 above)
Comment (BugenZhao) [reply src/common/src/array/stream_chunk_builder.rs:146]: Yes. There're no changes.

File: src/common/src/array/data_chunk.rs [hunk 6]
```diff
@@ -322,83 +320,24 @@ impl DataChunk {
...
-                    .collect();
-
-                let data_chunk = DataChunk::new(new_columns, array_len);
-                new_chunks.push(data_chunk);
-
-                new_chunk_require = std::cmp::min(total_capacity, each_size_limit);
-                array_len = new_chunk_require;
+        for chunk in chunks {
+            for output in builder.append_chunk(chunk.clone()) {
+                outputs.push(output);
```
Comment (stdrc) [inline src/common/src/array/data_chunk.rs:333]: Nice!

File: src/common/src/array/stream_chunk.rs (hunk 3 above)
Comment (stdrc) [inline src/common/src/array/stream_chunk.rs:142]: Ditto

**Review by stdrc — APPROVED** [review]
LGTM!

**Review by chenzl25 — APPROVED** [review]
LGTM
